content/blog/2023-08-03-parsing-ssh-config.markdown

```diff
@@ -0,0 +1,229 @@
+that the integrated terminal in VSCodium has no problem connecting to the
+remote. So the extension seems to be doing something different than just a plain
+SSH connection.
+
+You might think that the extension is not reading the SSH config. But the
+extension panel lists all the host aliases I've declared in the config, so it's
+clearly aware of the config at least. Possibly it doesn't understand the proxy
+config correctly? If it was trying to connect directly from the host, it would
+make sense to fail a DNS lookup.
+
+
+## Investigating
+
+Enough theorising, time to debug the extension as it tries to connect.
+
+From the error above, the string `"Error resolving authority"` looks like
+something I can search for. This takes me to the
+[`catch` case for a large try-catch block](https://github.com/jeanp413/open-remote-ssh/blob/521098e24f48b4b9e04d476895f9097b03f8c984/src/authResolver.ts#L226).
+It could be annoying to narrow down which part of the block
+throws the exception, but fortunately debugging is as easy as installing the
+dependencies and running the pre-configured 'Extension' debug target. This opens
+a new window with the local copy of the extension active, and I can debug it in
+the original window.
+
+In this block, there is a conditional statement on whether the `ProxyJump` field
+is present in the config. This is a good place to break on and see what the
+computed config looks like. If it doesn't find a proxy then of course it's going
+to run everything on the host.
+
+And indeed, it doesn't think there is a proxy. This is progress, but why does
+the extension's view of the config not match up with what SSH does? After all,
+invoking SSH directly connects properly. Tracing back the source of the config
+in the extension, it ultimately comes from manually reading in and parsing the
+SSH config. When resolving the host argument it manually computes the config as
+per [`ssh_config(5)`](https://man7.org/linux/man-pages/man5/ssh_config.5.html).
+Yet somewhere it makes a mistake, because it doesn't include the `ProxyJump`
+field.
+
+
+## Parsing SSH config
+
+To get to the bottom of this, we need to know the rules behind parsing SSH
+configs. The `ssh_config(5)` manpage does a pretty decent job of explaining
+this, but I'm going to go over the relevant information here. I reckon most
+people have a vague idea of how it works, and can write enough to meet their
+needs, but have never looked deeper into the actual rules behind how SSH parses
+the config.
+
+1. For starters, the config is parsed line by line. Leading whitespace (i.e.,
+   indentation) is ignored. So, while indentation makes it look like you are
+   configuring properties for a particular host, this isn't quite correct.
+   Instead, the `Host` and `Match` lines are special statements that enable or
+   disable all subsequent lines until the next `Host` or `Match`.
+
+    There is no backtracking; previous conditions and lines are not re-evaluated
```

Comment:
why is this indented wrong. idk if it changes the output, why is it wrong

Reply:
4 spaces required to be part of the bullet point. The indentation of the first paragraph was automatically done by the hard wrapper

Reply:
huh ok